jax/random.py

```diff
@@ -539,17 +539,18 @@ def _gamma_one(key, alpha):
   # https://en.wikipedia.org/wiki/Gamma_distribution#Generating_gamma-distributed_random_variables
   zero = _constant_like(alpha, 0)
   one = _constant_like(alpha, 1)
+  minus_one = _constant_like(alpha, -1)
   one_over_two = _constant_like(alpha, 0.5)
   one_over_three = _constant_like(alpha, 1. / 3.)
   squeeze_const = _constant_like(alpha, 0.0331)
   dtype = lax.dtype(alpha)
 
-  key, subkey = split(key)
   # for alpha < 1, we boost alpha to alpha + 1 and get a sample according to
   # Gamma(alpha) ~ Gamma(alpha+1) * Uniform()^(1 / alpha)
   boost = lax.select(lax.ge(alpha, one),
                      one,
-                     lax.pow(uniform(subkey, (), dtype=dtype), lax.div(one, alpha)))
+                     lax.pow(uniform(key, (), dtype=dtype), lax.div(one, alpha)))
+  key, = split(key, 1)
```

Comment:
I think this line should be reverted to the `key` / `subkey` pattern. In principle the callee could split the key itself and thus get the same key used here by the caller. Once a key is passed into a subroutine, it should be treated as unusable. (In this case, the key is being passed to what we know is an elementary random function, but even those might need to split keys, e.g. to generate more than 2^32 random values.)

Reply:
I agree that it is better to use `key` / `subkey` pattern so I have changed back to that pattern. But the compiling time will be slow as before.

Reply:
Hm I see your point now. I'll follow up with internal folks about the compile times; maybe there is some low-hanging fruit here.

Reply:
I don't want to force you to live with bad compile times. How about a temporary solution: let's go back to your original solution for now, but also leave a clear warning/todo comment on this line and [also in `random_bits`](https://github.com/google/jax/blob/c519d561b6f93ee355b5f5544822f6c06ecf1368/jax/random.py#L207) pointing out that `random.gamma` relies on the fact that `random.uniform` doesn't split the key. Then we can follow up with XLA folks to try to improve compile times so that we don't need this restriction. (We also plan to add some kind of checking for key reuse at some point.)

Is that plan better? We can either do that (i.e. add warning comments and go back to the "sequential keys" trick for now), or merge this PR as-is. In either case we'll follow up with XLA folks about this compilation benchmark. I defer to your judgement on which is best.

Reply:
Thanks @mattjj for pointing out the issue and following up! You are right that this is a good compilation benchmark. :) I think that it is better to keep it correct as-is and track down the performance in the future. If I change back to the original solution, then it'll be harder to decide when I should look at this issue again.

Reply:
Good plan. Thanks as always for the excellent contributions, and thanks for your patience on this PR. I'll merge this now, and also ping the PR thread when I get the XLA bug filed.